src/main/java/org/opensearch/neuralsearch/query/NeuralKNNQueryBuilder.java

```diff
@@ -0,0 +1,158 @@
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryRewriteContext;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.knn.index.query.KNNQueryBuilder;
+import org.opensearch.knn.index.query.rescore.RescoreContext;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * NeuralKNNQueryBuilder wraps KNNQueryBuilder to:
+ * 1. Isolate KNN plugin API changes to a single location
+ * 2. Allow extension with neural-search-specific information (e.g., query text)
+ */
+public class NeuralKNNQueryBuilder extends AbstractQueryBuilder<NeuralKNNQueryBuilder> {
+    private final KNNQueryBuilder knnQueryBuilder;
+
+    /**
+     * Wraps KNN Lucene query to support neural search extensions.
+     * Delegates core operations to the underlying KNN query.
+     */
+    public static class NeuralKNNQuery extends Query {
+        private final Query knnQuery;
+
+        public NeuralKNNQuery(Query knnQuery) {
+            this.knnQuery = knnQuery;
+        }
+
+        @Override
+        public String toString(String field) {
+            return knnQuery.toString(field);
+        }
+
+        @Override
+        public void visit(QueryVisitor visitor) {
+            // Delegate the visitor to the underlying KNN query
+            knnQuery.visit(visitor);
+        }
+
+        @Override
+        public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+            // Delegate weight creation to the underlying KNN query
+            return knnQuery.createWeight(searcher, scoreMode, boost);
+        }
+
+        @Override
+        public Query rewrite(IndexReader reader) throws IOException {
+            Query rewritten = knnQuery.rewrite(reader);
+            if (rewritten == knnQuery) {
+                return this;
+            }
+            return new NeuralKNNQuery(rewritten);
+        }
+
+        @Override
+        public boolean equals(Object other) {
+            if (this == other) return true;
+            if (other == null || getClass() != other.getClass()) return false;
+            NeuralKNNQuery that = (NeuralKNNQuery) other;
+            return Objects.equals(knnQuery, that.knnQuery);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(knnQuery);
+        }
+    }
+
+    /**
+     * Factory method to create NeuralKNNQueryBuilder with given parameters.
+     * Centralizes KNNQueryBuilder creation to handle breaking changes.
+     */
+    public static NeuralKNNQueryBuilder fromKNNBuilder(
```

Comment:
I think builder pattern might be better than having a method with large list of parameters. Also, the method name `fromKNNBuilder` is exposing internal implementation. We can simply call it `createNeuralKNNQueryBuilder()` but I think having builder pattern would be better than having this method.

Reply:
Updated to use builder pattern in NeuralKNNQueryBuilder.